File: hugegraph-server/hugegraph-kvt/src/main/native/KVTPushdownFunctions.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string>
#include <vector>
#include <map>
#include <unordered_set>
#include <algorithm>
#include <numeric>
#include <random>
#include <sstream>
#include "../../../kvt/kvt_inc.h"

// Helper function to decode variable-length integer (matching HugeGraph's format)
size_t decodeVInt(const unsigned char* data, size_t& bytes_read) {
    unsigned char leading = data[0];
    size_t value = leading & 0x7F;
    bytes_read = 1;
    
    if ((leading & 0x80) == 0) {
        return value;
    }
    
    for (int i = 1; i < 5; i++) {
        unsigned char b = data[bytes_read];
        bytes_read++;
        
        if ((b & 0x80) == 0) {
            value = (value << 7) | b;
            return value;
        } else {
            value = (value << 7) | (b & 0x7F);
        }
    }
    
    throw std::runtime_error("Invalid vInt encoding");
}

// Helper function to encode variable-length integer
void encodeVInt(size_t value, std::string& output) {
    if (value > 0x0fffffff) {
        output.push_back(0x80 | ((value >> 28) & 0x7f));
    }
    if (value > 0x1fffff) {
        output.push_back(0x80 | ((value >> 21) & 0x7f));
    }
    if (value > 0x3fff) {
        output.push_back(0x80 | ((value >> 14) & 0x7f));
    }
    if (value > 0x7f) {
        output.push_back(0x80 | ((value >> 7) & 0x7f));
    }
    output.push_back(value & 0x7f);
}

/**
 * Property filter function - filters entries based on property conditions
 */
bool hg_property_filter(KVTProcessInput& input, KVTProcessOutput& output) {
    if (!input.value || !input.parameter) {
        output.return_value = "Missing required input";
        return false;
    }
    
    const std::string& value = *input.value;
    const std::string& parameter = *input.parameter;
    
    try {
        // Parse parameter to get filter conditions
        const unsigned char* param_data = reinterpret_cast<const unsigned char*>(parameter.data());
        size_t param_pos = 0;
        size_t bytes_read = 0;
        
        // Read number of conditions
        size_t num_conditions = decodeVInt(param_data + param_pos, bytes_read);
        param_pos += bytes_read;
        
        // For simplicity, if all conditions match, keep the entry
        bool all_match = true;
        
        for (size_t i = 0; i < num_conditions && all_match; i++) {
            // Read property key
            size_t key_len = decodeVInt(param_data + param_pos, bytes_read);
            param_pos += bytes_read;
            std::string prop_key(parameter.data() + param_pos, key_len);
            param_pos += key_len;
            
            // Read relation type
            uint8_t relation = param_data[param_pos++];
            
            // Read value
            size_t val_len = decodeVInt(param_data + param_pos, bytes_read);
            param_pos += bytes_read;
            std::string filter_value(parameter.data() + param_pos, val_len);
            param_pos += val_len;
            
            // TODO: Implement actual property matching logic
            // For now, we'll just check if the value contains the filter string
            if (value.find(filter_value) == std::string::npos) {
                all_match = false;
            }
        }
        
        if (all_match) {
            // Keep the entry unchanged
            output.return_value = value;
            output.delete_key = false;
        } else {
            // Filter out the entry
            output.delete_key = false; // Don't actually delete, just don't return
            // Return empty to indicate filtered out
        }
        
        return true;
        
    } catch (const std::exception& e) {
        output.return_value = std::string("Filter error: ") + e.what();
        return false;
    }
}

/**
 * Count aggregation function - counts matching entries
 */
bool hg_count_aggregation(KVTProcessInput& input, KVTProcessOutput& output) {
    static uint64_t count = 0;
    static std::unordered_set<std::string> seen_keys;
    
    if (!input.parameter) {
        output.return_value = "Missing parameter";
        return false;
    }
    
    const std::string& parameter = *input.parameter;
    
    try {
        // Parse parameter
        const unsigned char* param_data = reinterpret_cast<const unsigned char*>(parameter.data());
        size_t param_pos = 0;
        
        // Read deduplicate flag
        bool deduplicate = param_data[param_pos++] != 0;
        
        // Initialize count on first entry
        if (input.range_first) {
            count = 0;
            seen_keys.clear();
        }
        
        // Count this entry
        if (deduplicate) {
            if (input.key && seen_keys.insert(*input.key).second) {
                count++;
            }
        } else {
            count++;
        }
        
        // Return count on last entry
        if (input.range_last) {
            output.return_value = std::to_string(count);
        }
        
        return true;
        
    } catch (const std::exception& e) {
        output.return_value = std::string("Count error: ") + e.what();
        return false;
    }
}

/**
 * Sum aggregation function - sums a property value across entries
 */
bool hg_sum_aggregation(KVTProcessInput& input, KVTProcessOutput& output) {
    static double sum = 0.0;
    
    if (!input.value || !input.parameter) {
        output.return_value = "Missing required input";
        return false;
    }
    
    try {
        // Initialize sum on first entry
        if (input.range_first) {
            sum = 0.0;
        }
        
        // Parse parameter to get property key
        const std::string& parameter = *input.parameter;
        const unsigned char* param_data = reinterpret_cast<const unsigned char*>(parameter.data());
        size_t param_pos = 0;
        size_t bytes_read = 0;
        
        // Read property key
        size_t key_len = decodeVInt(param_data + param_pos, bytes_read);
        param_pos += bytes_read;
        std::string prop_key(parameter.data() + param_pos, key_len);
        
        // TODO: Extract actual property value from the entry
        // For now, use a dummy value
        double value = 1.0;
        sum += value;
        
        // Return sum on last entry
        if (input.range_last) {
            output.return_value = std::to_string(sum);
        }
        
        return true;
        
    } catch (const std::exception& e) {
        output.return_value = std::string("Sum error: ") + e.what();
        return false;
    }
}

/**
 * Min/Max aggregation function
 */
bool hg_minmax_aggregation(KVTProcessInput& input, KVTProcessOutput& output) {
    static double extreme_value;
    static bool first_value;
    
    if (!input.value || !input.parameter) {
        output.return_value = "Missing required input";
        return false;
    }
    
    try {
        // Parse parameter
        const std::string& parameter = *input.parameter;
        const unsigned char* param_data = reinterpret_cast<const unsigned char*>(parameter.data());
        size_t param_pos = 0;
        
        // Read find_max flag
        bool find_max = param_data[param_pos++] != 0;
        
        // Initialize on first entry
        if (input.range_first) {
            first_value = true;
        }
        
        // TODO: Extract actual property value from the entry
        // For now, use a dummy value
        double value = 1.0;
        
        if (first_value) {
            extreme_value = value;
            first_value = false;
        } else {
            if (find_max) {
                extreme_value = std::max(extreme_value, value);
            } else {
                extreme_value = std::min(extreme_value, value);
            }
        }
        
        // Return result on last entry
        if (input.range_last) {
            output.return_value = std::to_string(extreme_value);
        }
        
        return true;
        
    } catch (const std::exception& e) {
        output.return_value = std::string("MinMax error: ") + e.what();
        return false;
    }
}

/**
 * Group-by aggregation function
 */
bool hg_groupby_aggregation(KVTProcessInput& input, KVTProcessOutput& output) {
    static std::map<std::string, double> groups;
    static std::map<std::string, uint64_t> counts; // For AVG calculation
    
    if (!input.value || !input.parameter) {
        output.return_value = "Missing required input";
        return false;
    }
    
    try {
        // Initialize on first entry
        if (input.range_first) {
            groups.clear();
            counts.clear();
        }
        
        // Parse parameter
        const std::string& parameter = *input.parameter;
        const unsigned char* param_data = reinterpret_cast<const unsigned char*>(parameter.data());
        size_t param_pos = 0;
        size_t bytes_read = 0;
        
        // Read group-by key
        size_t group_key_len = decodeVInt(param_data + param_pos, bytes_read);
        param_pos += bytes_read;
        std::string group_key(parameter.data() + param_pos, group_key_len);
        param_pos += group_key_len;
        
        // Read aggregation type
        uint8_t agg_type = param_data[param_pos++];
        
        // TODO: Extract actual group value and aggregate value from entry
        // For now, use dummy values
        std::string group_value = "group1";
        double agg_value = 1.0;
        
        // Perform aggregation based on type
        switch (agg_type) {
            case 0: // COUNT
                groups[group_value] += 1.0;
                break;
            case 1: // SUM
                groups[group_value] += agg_value;
                break;
            case 2: // AVG
                groups[group_value] += agg_value;
                counts[group_value]++;
                break;
            case 3: // MIN
                if (groups.find(group_value) == groups.end()) {
                    groups[group_value] = agg_value;
                } else {
                    groups[group_value] = std::min(groups[group_value], agg_value);
                }
                break;
            case 4: // MAX
                if (groups.find(group_value) == groups.end()) {
                    groups[group_value] = agg_value;
                } else {
                    groups[group_value] = std::max(groups[group_value], agg_value);
                }
                break;
        }
        
        // Return results on last entry
        if (input.range_last) {
            std::stringstream result;
            result << "{";
            bool first = true;
            for (const auto& [key, value] : groups) {
                if (!first) result << ",";
                first = false;
                
                double final_value = value;
                if (agg_type == 2 && counts[key] > 0) { // AVG
                    final_value = value / counts[key];
                }
                
                result << "\"" << key << "\":" << final_value;
            }
            result << "}";
            output.return_value = result.str();
        }
        
        return true;
        
    } catch (const std::exception& e) {
        output.return_value = std::string("GroupBy error: ") + e.what();
        return false;
    }
}

/**
 * Top-K function - returns top K entries based on a property
 */
bool hg_topk_function(KVTProcessInput& input, KVTProcessOutput& output) {
    static std::vector<std::pair<double, std::string>> top_entries;
    
    if (!input.value || !input.parameter) {
        output.return_value = "Missing required input";
        return false;
    }
    
    try {
        // Parse parameter
        const std::string& parameter = *input.parameter;
        const unsigned char* param_data = reinterpret_cast<const unsigned char*>(parameter.data());
        size_t param_pos = 0;
        size_t bytes_read = 0;
        
        // Read sort key
        size_t sort_key_len = decodeVInt(param_data + param_pos, bytes_read);
        param_pos += bytes_read;
        std::string sort_key(parameter.data() + param_pos, sort_key_len);
        param_pos += sort_key_len;
        
        // Read K value
        size_t k = decodeVInt(param_data + param_pos, bytes_read);
        param_pos += bytes_read;
        
        // Read ascending flag
        bool ascending = param_data[param_pos++] != 0;
        
        // Initialize on first entry
        if (input.range_first) {
            top_entries.clear();
            top_entries.reserve(k);
        }
        
        // TODO: Extract actual sort value from entry
        // For now, use a dummy value
        double sort_value = 1.0;
        
        // Add to top entries
        if (top_entries.size() < k) {
            top_entries.emplace_back(sort_value, *input.value);
        } else {
            // Check if this entry should replace the worst entry
            auto compare = ascending ? 
                [](const auto& a, const auto& b) { return a.first < b.first; } :
                [](const auto& a, const auto& b) { return a.first > b.first; };
            
            std::sort(top_entries.begin(), top_entries.end(), compare);
            
            if ((ascending && sort_value < top_entries.back().first) ||
                (!ascending && sort_value > top_entries.back().first)) {
                top_entries.back() = {sort_value, *input.value};
            }
        }
        
        // Return top K on last entry
        if (input.range_last) {
            auto compare = ascending ? 
                [](const auto& a, const auto& b) { return a.first < b.first; } :
                [](const auto& a, const auto& b) { return a.first > b.first; };
            
            std::sort(top_entries.begin(), top_entries.end(), compare);
            
            std::stringstream result;
            result << "[";
            for (size_t i = 0; i < top_entries.size(); i++) {
                if (i > 0) result << ",";
                result << "{\"value\":" << top_entries[i].first << ",\"data\":\"" 
                      << top_entries[i].second << "\"}";
            }
            result << "]";
            output.return_value = result.str();
        }
        
        return true;
        
    } catch (const std::exception& e) {
        output.return_value = std::string("TopK error: ") + e.what();
        return false;
    }
}

/**
 * Sampling function - randomly samples entries
 */
bool hg_sampling_function(KVTProcessInput& input, KVTProcessOutput& output) {
    static std::mt19937_64 rng;
    static std::uniform_real_distribution<double> dist(0.0, 1.0);
    static bool initialized = false;
    
    if (!input.value || !input.parameter) {
        output.return_value = "Missing required input";
        return false;
    }
    
    try {
        // Parse parameter
        const std::string& parameter = *input.parameter;
        const unsigned char* param_data = reinterpret_cast<const unsigned char*>(parameter.data());
        size_t param_pos = 0;
        
        // Read sample rate
        double sample_rate;
        memcpy(&sample_rate, param_data + param_pos, sizeof(double));
        param_pos += sizeof(double);
        
        // Read seed
        uint64_t seed;
        memcpy(&seed, param_data + param_pos, sizeof(uint64_t));
        param_pos += sizeof(uint64_t);
        
        // Initialize RNG on first entry
        if (input.range_first || !initialized) {
            rng.seed(seed);
            initialized = true;
        }
        
        // Decide whether to sample this entry
        if (dist(rng) < sample_rate) {
            output.return_value = *input.value;
        }
        // Else return nothing (filtered out)
        
        return true;
        
    } catch (const std::exception& e) {
        output.return_value = std::string("Sampling error: ") + e.what();
        return false;
    }
}

File: hugegraph-server/hugegraph-kvt/src/main/java/org/apache/hugegraph/backend/store/kvt/KVTPushdownFunctions.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.hugegraph.backend.store.kvt;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

import org.apache.hugegraph.backend.query.Condition;
import org.apache.hugegraph.backend.query.ConditionQuery;
import org.apache.hugegraph.backend.query.Query;
import org.apache.hugegraph.backend.serializer.BytesBuffer;
import org.apache.hugegraph.type.HugeType;
import org.apache.hugegraph.type.define.HugeKeys;
import org.apache.hugegraph.backend.id.Id;
import org.apache.hugegraph.util.E;
import org.apache.hugegraph.util.Log;
import org.slf4j.Logger;

/**
 * KVT Pushdown Functions for computation pushdown to the storage layer.
 * These functions are used with kvt_process and kvt_range_process to
 * perform filtering, aggregation, and other operations at the storage level.
 */
public class KVTPushdownFunctions {
    
    private static final Logger LOG = Log.logger(KVTPushdownFunctions.class);
    
    /**
     * Property filter pushdown for vertices and edges.
     * This function filters entries based on property conditions.
     */
    public static class PropertyFilterFunction {
        private final List<Condition> conditions;
        private final HugeType type;
        
        public PropertyFilterFunction(List<Condition> conditions, HugeType type) {
            this.conditions = conditions;
            this.type = type;
        }
        
        /**
         * Build parameter bytes for the filter function.
         * Format: [num_conditions][condition1][condition2]...
         * Each condition: [key_len][key][relation][value_len][value]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write number of conditions
            buffer.writeVInt(conditions.size());
            
            for (Condition condition : conditions) {
                if (!(condition instanceof Condition.Relation)) {
                    continue; // Skip non-relation conditions
                }
                Condition.Relation relation = (Condition.Relation) condition;
                
                // Write property key
                String key = relation.key().toString();
                buffer.writeVInt(key.length());
                buffer.writeBytes(key.getBytes());
                
                // Write relation type
                buffer.write((byte) relation.relation().ordinal());
                
                // Write value
                Object value = relation.value();
                byte[] valueBytes = serializeValue(value);
                buffer.writeVInt(valueBytes.length);
                buffer.writeBytes(valueBytes);
            }
            
            return buffer.bytes();
        }
        
        private byte[] serializeValue(Object value) {
            BytesBuffer buffer = BytesBuffer.allocate(256);
            if (value instanceof String) {
                buffer.write((byte) 0); // String type
                buffer.writeString((String) value);
            } else if (value instanceof Number) {
                buffer.write((byte) 1); // Number type
                buffer.writeDouble(((Number) value).doubleValue());
            } else if (value instanceof Boolean) {
                buffer.write((byte) 2); // Boolean type
                buffer.writeBoolean((Boolean) value);
            } else if (value instanceof Id) {
                buffer.write((byte) 3); // Id type
                buffer.writeId((Id) value);
            } else {
                // Default to string representation
                buffer.write((byte) 0);
                buffer.writeString(value.toString());
            }
            return buffer.bytes();
        }
    }
    
    /**
     * Count aggregation pushdown.
     * This function counts matching entries during a range scan.
     */
    public static class CountAggregationFunction {
        private final List<Condition> filterConditions;
        private final boolean deduplicate;
        
        public CountAggregationFunction(List<Condition> filterConditions, 
                                       boolean deduplicate) {
            this.filterConditions = filterConditions;
            this.deduplicate = deduplicate;
        }
        
        /**
         * Build parameter bytes for count aggregation.
         * Format: [deduplicate_flag][num_conditions][conditions...]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write deduplicate flag
            buffer.writeBoolean(deduplicate);
            
            // Write filter conditions if any
            if (filterConditions != null && !filterConditions.isEmpty()) {
                PropertyFilterFunction filter = new PropertyFilterFunction(
                    filterConditions, HugeType.VERTEX);
                byte[] filterParams = filter.buildParameter();
                buffer.writeBytes(filterParams);
            } else {
                buffer.writeVInt(0); // No conditions
            }
            
            return buffer.bytes();
        }
    }
    
    /**
     * Sum aggregation pushdown.
     * This function sums a specific property value during a range scan.
     */
    public static class SumAggregationFunction {
        private final String propertyKey;
        private final List<Condition> filterConditions;
        
        public SumAggregationFunction(String propertyKey, 
                                     List<Condition> filterConditions) {
            this.propertyKey = propertyKey;
            this.filterConditions = filterConditions;
        }
        
        /**
         * Build parameter bytes for sum aggregation.
         * Format: [property_key_len][property_key][num_conditions][conditions...]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write property key to sum
            buffer.writeVInt(propertyKey.length());
            buffer.writeBytes(propertyKey.getBytes());
            
            // Write filter conditions if any
            if (filterConditions != null && !filterConditions.isEmpty()) {
                PropertyFilterFunction filter = new PropertyFilterFunction(
                    filterConditions, HugeType.VERTEX);
                byte[] filterParams = filter.buildParameter();
                buffer.writeBytes(filterParams);
            } else {
                buffer.writeVInt(0); // No conditions
            }
            
            return buffer.bytes();
        }
    }
    
    /**
     * Min/Max aggregation pushdown.
     * This function finds the minimum or maximum value of a property.
     */
    public static class MinMaxAggregationFunction {
        private final String propertyKey;
        private final boolean findMax; // true for max, false for min
        private final List<Condition> filterConditions;
        
        public MinMaxAggregationFunction(String propertyKey, 
                                        boolean findMax,
                                        List<Condition> filterConditions) {
            this.propertyKey = propertyKey;
            this.findMax = findMax;
            this.filterConditions = filterConditions;
        }
        
        /**
         * Build parameter bytes for min/max aggregation.
         * Format: [find_max_flag][property_key_len][property_key][conditions...]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write find max flag
            buffer.writeBoolean(findMax);
            
            // Write property key
            buffer.writeVInt(propertyKey.length());
            buffer.writeBytes(propertyKey.getBytes());
            
            // Write filter conditions if any
            if (filterConditions != null && !filterConditions.isEmpty()) {
                PropertyFilterFunction filter = new PropertyFilterFunction(
                    filterConditions, HugeType.VERTEX);
                byte[] filterParams = filter.buildParameter();
                buffer.writeBytes(filterParams);
            } else {
                buffer.writeVInt(0); // No conditions
            }
            
            return buffer.bytes();
        }
    }
    
    /**
     * Group-by aggregation pushdown.
     * This function groups entries by a property and performs aggregation.
     */
    public static class GroupByAggregationFunction {
        private final String groupByKey;
        private final String aggregateKey;
        private final AggregationType aggregationType;
        private final List<Condition> filterConditions;
        
        public enum AggregationType {
            COUNT,
            SUM,
            AVG,
            MIN,
            MAX
        }
        
        public GroupByAggregationFunction(String groupByKey,
                                         String aggregateKey,
                                         AggregationType aggregationType,
                                         List<Condition> filterConditions) {
            this.groupByKey = groupByKey;
            this.aggregateKey = aggregateKey;
            this.aggregationType = aggregationType;
            this.filterConditions = filterConditions;
        }
        
        /**
         * Build parameter bytes for group-by aggregation.
         * Format: [group_key_len][group_key][agg_type][agg_key_len][agg_key][conditions...]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write group-by key
            buffer.writeVInt(groupByKey.length());
            buffer.writeBytes(groupByKey.getBytes());
            
            // Write aggregation type
            buffer.write((byte) aggregationType.ordinal());
            
            // Write aggregate key (may be empty for COUNT)
            if (aggregateKey != null) {
                buffer.writeVInt(aggregateKey.length());
                buffer.writeBytes(aggregateKey.getBytes());
            } else {
                buffer.writeVInt(0);
            }
            
            // Write filter conditions if any
            if (filterConditions != null && !filterConditions.isEmpty()) {
                PropertyFilterFunction filter = new PropertyFilterFunction(
                    filterConditions, HugeType.VERTEX);
                byte[] filterParams = filter.buildParameter();
                buffer.writeBytes(filterParams);
            } else {
                buffer.writeVInt(0); // No conditions
            }
            
            return buffer.bytes();
        }
    }
    
    /**
     * Top-K pushdown.
     * This function returns the top K entries based on a property value.
     */
    public static class TopKFunction {
        private final String sortKey;
        private final int k;
        private final boolean ascending;
        private final List<Condition> filterConditions;
        
        public TopKFunction(String sortKey, int k, boolean ascending,
                           List<Condition> filterConditions) {
            this.sortKey = sortKey;
            this.k = k;
            this.ascending = ascending;
            this.filterConditions = filterConditions;
        }
        
        /**
         * Build parameter bytes for top-k function.
         * Format: [sort_key_len][sort_key][k][ascending_flag][conditions...]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write sort key
            buffer.writeVInt(sortKey.length());
            buffer.writeBytes(sortKey.getBytes());
            
            // Write k value
            buffer.writeVInt(k);
            
            // Write ascending flag
            buffer.writeBoolean(ascending);
            
            // Write filter conditions if any
            if (filterConditions != null && !filterConditions.isEmpty()) {
                PropertyFilterFunction filter = new PropertyFilterFunction(
                    filterConditions, HugeType.VERTEX);
                byte[] filterParams = filter.buildParameter();
                buffer.writeBytes(filterParams);
            } else {
                buffer.writeVInt(0); // No conditions
            }
            
            return buffer.bytes();
        }
    }
    
    /**
     * Distinct/Deduplicate pushdown.
     * This function removes duplicate entries based on specified keys.
     */
    public static class DistinctFunction {
        private final List<String> distinctKeys;
        private final List<Condition> filterConditions;
        
        public DistinctFunction(List<String> distinctKeys,
                               List<Condition> filterConditions) {
            this.distinctKeys = distinctKeys;
            this.filterConditions = filterConditions;
        }
        
        /**
         * Build parameter bytes for distinct function.
         * Format: [num_keys][key1_len][key1][key2_len][key2]...[conditions...]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write number of distinct keys
            buffer.writeVInt(distinctKeys.size());
            
            // Write each distinct key
            for (String key : distinctKeys) {
                buffer.writeVInt(key.length());
                buffer.writeBytes(key.getBytes());
            }
            
            // Write filter conditions if any
            if (filterConditions != null && !filterConditions.isEmpty()) {
                PropertyFilterFunction filter = new PropertyFilterFunction(
                    filterConditions, HugeType.VERTEX);
                byte[] filterParams = filter.buildParameter();
                buffer.writeBytes(filterParams);
            } else {
                buffer.writeVInt(0); // No conditions
            }
            
            return buffer.bytes();
        }
    }
    
    /**
     * Sampling pushdown.
     * This function samples entries randomly or systematically.
     */
    public static class SamplingFunction {
        private final double sampleRate;
        private final long seed;
        private final List<Condition> filterConditions;
        
        public SamplingFunction(double sampleRate, long seed,
                               List<Condition> filterConditions) {
            E.checkArgument(sampleRate > 0 && sampleRate <= 1.0,
                           "Sample rate must be between 0 and 1");
            this.sampleRate = sampleRate;
            this.seed = seed;
            this.filterConditions = filterConditions;
        }
        
        /**
         * Build parameter bytes for sampling function.
         * Format: [sample_rate][seed][conditions...]
         */
        public byte[] buildParameter() {
            BytesBuffer buffer = BytesBuffer.allocate(1024);
            
            // Write sample rate
            buffer.writeDouble(sampleRate);
            
            // Write random seed
            buffer.writeLong(seed);
            
            // Write filter conditions if any
            if (filterConditions != null && !filterConditions.isEmpty()) {
                PropertyFilterFunction filter = new PropertyFilterFunction(
                    filterConditions, HugeType.VERTEX);
                byte[] filterParams = filter.buildParameter();
                buffer.writeBytes(filterParams);
            } else {
                buffer.writeVInt(0); // No conditions
            }
            
            return buffer.bytes();
        }
    }
}